// pose-viewer/dist/collection/components/pose-viewer/renderers/interactive.pose-renderer.js
import { PoseRenderer } from "./pose-renderer";
import { h } from "@stencil/core";
import * as THREE from "three";
import { OrbitControls } from "three/examples/jsm/controls/OrbitControls.js";
export class InteractivePoseRenderer extends PoseRenderer {
    constructor() {
        super(...arguments);
        // Reuse the same geometry for all joints.
        this.jointGeometry = new THREE.SphereGeometry(2, 16, 16);
        // Continuous render loop.
        this.animate = () => {
            requestAnimationFrame(this.animate);
            this.controls.update();
            this.renderer.render(this.scene, this.camera);
        };
    }
    // Convert a joint to a 3D point.
    transform(joint) {
        const { width, height } = this.viewer.pose.header;
        const x = joint.X - width / 2;
        const y = height / 2 - joint.Y;
        const z = joint.Z;
        return new THREE.Vector3(x, y, z);
    }
    // Create a mesh for a joint.
    renderJoint(_, joint, color) {
        const material = new THREE.MeshBasicMaterial({
            color: new THREE.Color(color.R / 255, color.G / 255, color.B / 255),
            transparent: true,
            opacity: joint.C,
        });
        const sphere = new THREE.Mesh(this.jointGeometry, material);
        sphere.position.copy(this.transform(joint));
        return sphere;
    }
    // Create a line between two joints.
    renderLimb(from, to, color) {
        const material = new THREE.LineBasicMaterial({
            color: new THREE.Color(color.R / 255, color.G / 255, color.B / 255),
            transparent: true,
            opacity: (from.C + to.C) / 2,
        });
        const points = [this.transform(from), this.transform(to)];
        const geometry = new THREE.BufferGeometry().setFromPoints(points);
        return new THREE.Line(geometry, material);
    }
    // Initialize Three.js components.
    initThree(canvas) {
        this.renderer = new THREE.WebGLRenderer({ canvas, antialias: true, alpha: true });
        this.renderer.setSize(this.viewer.elWidth, this.viewer.elHeight);
        this.scene = new THREE.Scene();
        if (this.viewer.background) {
            this.scene.background = new THREE.Color(this.viewer.background);
        }
        this.camera = new THREE.PerspectiveCamera(75, this.viewer.elWidth / this.viewer.elHeight, 0.1, 1000);
        const size = Math.max(this.viewer.pose.header.width, this.viewer.pose.header.height);
        this.camera.position.set(0, 0, size);
        this.controls = new OrbitControls(this.camera, canvas);
        this.controls.enableDamping = true;
        const ambientLight = new THREE.AmbientLight(0xffffff, 0.8);
        this.scene.add(ambientLight);
        this.poseGroup = new THREE.Group();
        this.scene.add(this.poseGroup);
        this.animate();
    }
    // Helper to dispose objects and free resources.
    disposeObject(obj) {
        if (obj instanceof THREE.Mesh) {
            // Avoid disposing shared geometry.
            if (obj.geometry !== this.jointGeometry) {
                obj.geometry.dispose();
            }
            if (Array.isArray(obj.material)) {
                obj.material.forEach(mat => mat.dispose());
            }
            else {
                obj.material.dispose();
            }
        }
        else if (obj instanceof THREE.Line) {
            obj.geometry.dispose();
            obj.material.dispose();
        }
    }
    // Update the scene with the latest pose frame.
    updateScene(frame) {
        // Dispose of previous objects to avoid memory leaks.
        this.poseGroup.children.forEach(child => this.disposeObject(child));
        this.poseGroup.clear();
        const rendered = this.renderFrame(frame);
        const objects = rendered.flat(Infinity).filter((o) => o instanceof THREE.Object3D);
        objects.forEach(obj => this.poseGroup.add(obj));
        // Update orbit controls target.
        this.poseGroup.updateMatrixWorld(true);
        const box = new THREE.Box3().setFromObject(this.poseGroup);
        const center = new THREE.Vector3();
        box.getCenter(center);
        this.controls.target.copy(center);
    }
    // Main render method.
    render(frame) {
        const initRenderer = () => {
            const canvas = this.viewer.element.shadowRoot.querySelector("canvas");
            if (canvas && !this.renderer) {
                this.initThree(canvas);
            }
            if (this.renderer) {
                this.updateScene(frame);
            }
            else {
                requestAnimationFrame(initRenderer);
            }
        };
        initRenderer();
        return (h("canvas", { width: this.viewer.elWidth, height: this.viewer.elHeight }));
    }
}
